### PR TITLE
How To API Evolution Article

### PR DESCRIPTION
**Pull request by philsturgeon**
Possibly needs more use cases, or better ones, or examples of things you cant do.

[Have a read!](https://github.com/philsturgeon/philsturgeon.github.io/blob/0ecde17/source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown)

### CODE REVIEW

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+JSON Schema is considering [adding a deprecated keyword](https://github.com/json-schema-org/json-schema-spec/issues/74), and oops I think I'm in charge of making that happen. I'll get back to doing that after this blog post. The idea here would be to pair the schema with a smart SDK (client code) which detects which fields are being used. If the schema marks the `foo` field as deprecated, and the client code then calls `$response->foo`, in PHP that could be snooped on with `function __get` (dynamic programming yaaaay) and fire off deprecation warnings on use.
+
+GraphQL has the advantage when it comes to field deprecation for sure, as their type system demands clients to specify the fields they want. This is the sort of advantage you get when your type system, clients, etc. are all part of the same package, but HTTP in general can achieve this same functionality through standards.
```
Comment (MikeRalphson) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: This comes across as a bit of a non-sequiteur - at least to me. How does having the client specify required fields **actually help** if you want to make one of those fields go away, or be deprecated?

Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Gotcha! Will expand. Basically if you know who's asking for it, you can contact that client (manually or automatically) and suggest they change it. Or shove warnings into the payload somehow.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+Now clients can either send the `price` field, and it'll convert. This will only work for the older currencies, and if `currency` is a field in the resource (or something nearby) then it's easy enough to support `price` for whatever initial currencies you had (dollar/pound/euro) and throw an error if somebody tries using price for these newer currencies, pointing them instead to the micro field.
+
+Nothing broke for existing use cases, and new functionality was added seamlessly.
+
+> We have too many old fields kicking around, we need to get rid of them.
+
+Deprecations can be communicated in a few ways for API's. For those using OpenAPI v3, you can mark it as `deprecated: true` in the documentation. That's no ideal, of course, as OpenAPI is usually human-readable documentation, sat out of band on a developer portal somewhere. Rarely are OpenAPI schemas shoved into the response like JSON Schema is, so programatically clients have no real way to access this.
```
Comment (adamaltman) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: typo: should be... `programmatically` (I think)

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+All of that said, removing old fields is usually not all that much of a rush or an issue. Over time new developers will be writing new integrations, looking at your new documentation that tells them to use the new field, and your developer newsletters or changelogs just let them know to move away from it over time.
+
+> A carpooling company that has "matches" as a relationship between "drivers" and "passengers", suggesting folks who could ride together, containing properties like "passenger_id" and "driver_id". Now we need multiple drivers, so this whole matches concept is garbage.
+
+At a lot of startups, this sort of conceptual change is common. No number of new fields is going to help out here, as the whole "one record = one match = one driver + one passenger" concept was junk. We'd need to make it so folks could accept a carpool based on the group, and any one of those folks could drive on a given day.
+
+Luckily, business names often change fairly regularly in the sort of companies that have fundamental changes lie this. There is often a better word that folks have been itching to switch to, and evolution gives you a chance to leverage that change to your benefit.
```
Comment (cebe) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: `lie` -> `like`

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+Deprecating the whole concept of "matches", a new concept of "riders" can be created. This resource would track folks far beyond just being "matched", through the whole lifecycle of the carpool, thanks to a status field containing pending, active, inactive, blocked, etc.
+
+By creating this a new `/riders` endpoint, this new resource can have a brand new representation. As always, the same database fields can be used internally, the same internal alerting tools are used for letting folks know about matches (v1 app) or new pending riders (v2 app). The API can create and update "matches" through this new "riders" interface. Clients can then use either one, and the code just figures itself out in the background. Over time the refactoring can be done to move the internal logic more towards riders, and your [integration tests / contract tests](https://blog.apisyouwonthate.com/tricking-colleagues-into-writing-documentation-via-contract-testing-206308b47e05) will confirm that things aren't changing on the outside.
+
```
Comment (cebe) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: how would the `matches` endpoint return data where there is more than one driver? If the data does not fit the endpoint anymore, it must be broken or fail for such data. Mentioning this would show the limits of evolution.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 2 above)
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Thanks, fixed!

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+```
+
+When folks `POST` or `PATCH` to your API, if they send a `name` you can convert it, if they send `first_name` and `last_name` it'll get picked up fine on the serializer. Job done.
+
+> The field `price` needs to stop being dollars/pounds/whatever as we're starting to support currencies that don't fit into "unit" and "subunit".
+
+Ok! [Currencies are hard](https://gist.github.com/rgs/6509585), so this sort of thing is not uncommon. Let's start [using micros](https://support.siftscience.com/hc/en-us/articles/203869406-The-amount-field) instead.
+
+We don't want to just outright change this value from dollars to cents because then we'd start charging $1,000,000 for stuff that should only cost $1, and folks probably wouldn't like that. Let's make a new field: `price_micros`.
+
```
Comment (fmvilas) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: I may be too strict here, but 1 dollar is not 1,000,000 cents. If you're just exaggerating, then forget my comment.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+Deprecating the whole concept of "matches", a new concept of "riders" can be created. This resource would track folks far beyond just being "matched", through the whole lifecycle of the carpool, thanks to a status field containing pending, active, inactive, blocked, etc.
+
+By creating this a new `/riders` endpoint, this new resource can have a brand new representation. As always, the same database fields can be used internally, the same internal alerting tools are used for letting folks know about matches (v1 app) or new pending riders (v2 app). The API can create and update "matches" through this new "riders" interface. Clients can then use either one, and the code just figures itself out in the background. Over time the refactoring can be done to move the internal logic more towards riders, and your [integration tests / contract tests](https://blog.apisyouwonthate.com/tricking-colleagues-into-writing-documentation-via-contract-testing-206308b47e05) will confirm that things aren't changing on the outside.
```
Comment (fmvilas) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: 3 times `new` in the first sentence. You might want to rephrase it.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+The above solutions are a little ad-hoc, and can lead to branched code paths with a bunch of ifs. You can feature flag some of this stuff to help keep things a little tidy, and another approach is to write up change as libraries, something [Stripe refer to as "declarative changes"](https://stripe.com/blog/api-versioning). This approach can be a little heavy handed, but it's something to keep in mind. 
+
+## Summary
+
+Evolution involves thinking a little differently to how you approach change. Often there are simple things you can do to keep clients ticking along, and whilst clients will have to change at _some point_, the whole goal here is to allow them a decent amount of time to make that switch, with the minimal change possible during that change, and **no lock-step deploys** required.
+
+And yes, whilst making a new endpoint to switch `/matches` and `/riders` is essentially the same as `/v1/matches` and `/v2/matches`, you've skipped the [quagmire of tradeoffs](https://blog.apisyouwonthate.com/api-versioning-has-no-right-way-f3c75457c0b7) between global versioning, resource versioning, or **gulp** method versioning. Global versioning has it's place, but so does evolution.
+
```
Comment (fmvilas) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Should it be "has its place"?

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+There are a lot of pro's and con's to various approaches to API versioning, but that has been covered in depth before: _[API Versioning Has No "Right" Way](https://blog.apisyouwonthate.com/api-versioning-has-no-right-way-f3c75457c0b7)_.
+
+API Evolution is making a comeback these days with GraphQL and gRPC advocates shouting about it. Whatever API paradigm or implementation you subscribe to, evolution is available to you. REST advocates have been recommending API evolution for decades, but in the past I failed to understand how exactly to handle evolution.
+
+Luckily, as always, tooling and standards for HTTP have been improving, and these days API evolution is a lot easier to wrap your head around.
+
+## What is API Evolution
+
+API Evolution is the concept of striving to maintain the "I" in API, the request/response body, query parameters, general functionality, etc., only breaking them when you absolutely, _absolutely_, have to. It's the idea that API developers bending over backwards to maintain a contract, not matter how annoying that might be, is often more financially and logistically viable than dumping onto a wide array of clients.
```
Comment (kevinswiber) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: `not matter how` -> `no matter how`

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+At some point change cannot be prevented, so at that time evolution suggests you provide sensible warnings to clients, letting them know if a feature they're using is going away, and not bothering them otherwise.
+
+## Examples
+
+> The field `name` exists, and that needs to be splint into `first_name` and `last_name`.
```
Comment (kevinswiber) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: `splint` -> `split`

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 5 above)
Comment (cjunge-work) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Phil's using microcents :)

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+
+Putting [client-side validation logic in JSON Schema](https://blog.apisyouwonthate.com/the-many-amazing-uses-of-json-schema-client-side-validation-c78a11fbde45
+), and having clients use that as the basis for their validation, solves this problem and a whoooole lot more. Over time new JSON Schema files can be pushed, and clients will pick those up from the `Link` header as they go, giving seamless support for the new validation rules!
+
+> Deprecating a specific type of authentication from an endpoint, it's time to say goodbye to HTTP Basic Auth.
+
+If they the client is making a request with an authorization header, they have some sort of account. If during the signup for that account you've asked them for an email, you can contact them. If you've not got any way to contact them... tough. Monitor how many folks are using HTTP basic, blog about it, shove some videos up, and eventually you're just going to have to turn it off.
```
Comment (kevinswiber) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: `If they the` -> `If the`

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 4 above)
Comment (nhocki) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: @philsturgeon If I remember correctly, we only exposed the first driver to the older apps. If the user accepted it, great... The other ones would be denied...

If the user rejected the first one, we would show the next one, and the next one, and the next one...

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 6 above)
Comment (josephgardner) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: typo, "creating ~~this~~ a new"

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+GraphQL has the advantage when it comes to field deprecation for sure, as their type system demands clients to specify the fields they want. By knowing which clients are requesting a deprecated field, you can either reach out to that client (manually or automatically), or shove some warnings into the response somewhere to let them know they're asking for a thing which is going away. This is the sort of advantage you get when your type system, clients, etc. are all part of the same package, but HTTP in general can achieve this same functionality through standards.
+
+All of that said, removing old fields is usually not all that much of a rush or an issue. Over time new developers will be writing new integrations, looking at your new documentation that tells them to use the new field, and your developer newsletters or changelogs just let them know to move away from it over time.
+
+> A carpooling company that has "matches" as a relationship between "drivers" and "passengers", suggesting folks who could ride together, containing properties like "passenger_id" and "driver_id". Now we need multiple drivers, so this whole matches concept is garbage.
```
Comment (josephgardner) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: > Now we need multiple drivers

This example is a bit hard to follow. Do you mean "now we need to support a person being both a driver OR a passenger?"

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+
+Clients then add a middleware to their HTTP calls, checking for Sunset headers. We do this with [faraday-sunset](https://github.com/wework/faraday-sunset) (Ruby), [Hunter Skrasek](https://twitter.com/HSkrasek/) made [guzzle-sunset](https://github.com/hskrasek/guzzle-sunset) (PHP), and anyone can write a thing that looks for a header and logs it to whatever logging thing they're using.
+
+[HTTP date]: https://tools.ietf.org/html/rfc7231#section-7.1.1.1
+[sunset-draft]: https://tools.ietf.org/html/draft-wilde-sunset-header-03
+[rails-sunset]: https://github.com/wework/rails-sunset/
+
+> Changing validation rules, something seemingly backwards compatible, like making the length of a field a little longer.
+
+Changing validation rules seems like it might be a backwards compatible change, but often it is not. For example, making a string field accept a longer value can lead to some interesting problems. As clients often bake validation rules into client applications based on whatever the documentation says, if that documentation changes they aren't going to notice, and that validation will mismatch on different clients.
```
Comment (josephgardner) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: This isn't very convincing as backwards incompatible; this particular example won't "break" any clients, just they won't benefit from the additional available length. Maybe in addition, provide an example of wanting to introduce more restrictive validation which is definitely breaking.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+## Summary
+
+Evolution involves thinking a little differently to how you approach change. Often there are simple things you can do to keep clients ticking along, and whilst clients will have to change at _some point_, the whole goal here is to allow them a decent amount of time to make that switch, with the minimal change possible during that change, and **no lock-step deploys** required.
+
+And yes, whilst making a new endpoint to switch `/matches` and `/riders` is essentially the same as `/v1/matches` and `/v2/matches`, you've skipped the [quagmire of tradeoffs](https://blog.apisyouwonthate.com/api-versioning-has-no-right-way-f3c75457c0b7) between global versioning, resource versioning, or **gulp** method versioning. Global versioning has it's place, but so does evolution.
+
+Think about it this way. If implementing some change takes twice as long for API developers compared to other versioning approaches, but save 6 or 7 client developer teams from having to do a while bunch of work, testing, etc. to chase new versions, this has been worthwhile to the company in terms of engineering hours spent.
```
Comment (josephgardner) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: typo "while bunch of work"

Comment (josephgardner) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: > worthwhile to the company in terms of engineering hours spent.

Also worth mentioning is the ROI for keeping customers/clients happy

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 12 above)
Comment (tpluscode) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: So, maybe this example could be change to a scenario where the maximum length is decreased?

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 7 above)
Comment (tpluscode) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Well, `/matches => /riders` is not the same as `/v1/matches => /v2/matches`. The latter affects the entire API space. I see where you're coming from. Maybe this statement can be softened a little. 

``` diff
-is essentially the same as
+is not much better than
```

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 4 above)
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Aha! That was it. Awesome work :)

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 5 above)
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Great spot! I changed from subunits to micros mid article and some of this crap was left over. I cleaned it up here. ffa7c69cb

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 6 above)
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Gross! Overedited sentance makes no sense. Fixed.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 7 above)
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: English is my _first_ language. 😅

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 8 above)
[the remark this answers is not in the document]
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Got it! Thank you: 

>  It's the idea that API developers bending over backwards to maintain a contract, no matter how annoying that might be, is often more financially and logistically viable than dumping the workload a wide array of clients.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    "#{object.first_name} #{object.last_name}"
+  end
+end
+```
+
+When folks `POST` or `PATCH` to your API, if they send a `name` you can convert it, if they send `first_name` and `last_name` it'll get picked up fine on the serializer. Job done.
+
+> The field `price` needs to stop being dollars/pounds/whatever as we're starting to support currencies that don't fit into "unit" and "subunit".
+
+Ok! [Currencies are hard](https://gist.github.com/rgs/6509585), so this sort of thing is not uncommon. Let's start [using micros](https://support.siftscience.com/hc/en-us/articles/203869406-The-amount-field) instead.
```
[the remark this answers is not in the document]
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: I expanded on the topic a bit. It's used in a fair few places, but basicalluy im kinda teach folks some things in the process. Otherwise people say "ugh thats wrong why would you recommend that" and I dont need that noise right now.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 12 above)
Comment (darrelmiller) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: @josephgardner  It can break clients.  Client libraries are often built to allow you to GET an object, make some modifications and then PUT it back.  If you are using a down-rev client and retrieve an object that has a description that is longer than the old constraint, the GET/change/PUT flow will fail.

Comment (darrelmiller) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: There is an even more insidious issue with optional fields which all assume are non-breaking.  However, what happens when an up-rev client sets an option value and then a down-rev client that knows nothing about the optional value, does a GET/change/PUT flow.  The update will not contain the optional value and the PUT semantics cause the optional value to be deleted.  So far the only solution I have found for this is to version the message.

Comment (darrelmiller) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: … or do user-agent sniffing  (ick)

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 11 above)
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: > Now we need to support carpools that can have multiple drivers (i.e: Frank and Sally both take it in turns to drive), so this whole matches concept is garbage.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 12 above)
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Ahh, well if you read the attached link it's absolutely a breaking change and will royally fuck things up.

Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Updated this in 06712a47d, thank you !

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+
+There are a lot of pro's and con's to various approaches to API versioning, but that has been covered in depth before: _[API Versioning Has No "Right" Way](https://blog.apisyouwonthate.com/api-versioning-has-no-right-way-f3c75457c0b7)_.
+
+API Evolution is making a comeback these days with GraphQL and gRPC advocates shouting about it. Whatever API paradigm or implementation you subscribe to, evolution is available to you. REST advocates have been recommending API evolution for decades, but in the past I failed to understand how exactly to handle evolution.
+
+Luckily, as always, tooling and standards for HTTP have been improving, and these days API evolution is a lot easier to wrap your head around.
+
+## What is API Evolution
+
+API Evolution is the concept of striving to maintain the "I" in API, the request/response body, query parameters, general functionality, etc., only breaking them when you absolutely, _absolutely_, have to. It's the idea that API developers bending over backwards to maintain a contract, no matter how annoying that might be, is often more financially and logistically viable than dumping the workload a wide array of clients.
```
Comment (tpluscode) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Should it be "dumping the workload *on*"?

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 16]
```diff
@@ -0,0 +1,146 @@
+---
+title: API Evolution for REST/HTTP APIs
+date: 2018-05-02 21:01 UTC
+tags:
+---
+
+There are a lot of pro's and con's to various approaches to API versioning, but that has been covered in depth before: _[API Versioning Has No "Right" Way](https://blog.apisyouwonthate.com/api-versioning-has-no-right-way-f3c75457c0b7)_.
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: *pros and cons (no apostrophes)

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 17]
```diff
@@ -0,0 +1,146 @@
+---
+title: API Evolution for REST/HTTP APIs
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Throughout this doc: is it API Evolution (capital E) or API evolution (lower case e?). I don't care either way as long as it's consistent.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+
+  attributes :name, :first_name, :last_name
+
+  attribute :name do |object|
+    "#{object.first_name} #{object.last_name}"
+  end
+end
+```
+
+When folks `POST` or `PATCH` to your API, if they send a `name` you can convert it, if they send `first_name` and `last_name` it'll get picked up fine on the serializer. Job done.
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: *or if they send

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 19]
```diff
@@ -0,0 +1,146 @@
...
+
+> The field `price` needs to stop being dollars/pounds/whatever as we're starting to support currencies that don't fit into "unit" and "subunit".
+
+Switching to an integer to place your cents, pence, etc. would be just as much of a [Fallacies Programmers Think About Currencies](https://gist.github.com/rgs/6509585) as using float dollars/pounds, etc. To support the widest array of currencies, some folks like to use "micros", a concept explained well here by [Sift Science](https://support.siftscience.com/hc/en-us/articles/203869406-The-amount-field). In this case, the new field could easily be called `price_micros`.
+
+Want to keep it simple and go with cents, pence, etc.? Fine, `price_subunits` will do. If somebody grumps about that and you want a more concise name, just call it `amount` and point folks towards that field instead. A thesaurus is handy.
+
+Why don't we just outright change this value from dollars to micros? because then we'd start charging $1,000,000 for stuff that should only cost $1, and folks probably wouldn't like that.
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: *Because

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 20]
```diff
@@ -0,0 +1,146 @@
...
+end
+```
+
+When folks `POST` or `PATCH` to your API, if they send a `name` you can convert it, if they send `first_name` and `last_name` it'll get picked up fine on the serializer. Job done.
+
+> The field `price` needs to stop being dollars/pounds/whatever as we're starting to support currencies that don't fit into "unit" and "subunit".
+
+Switching to an integer to place your cents, pence, etc. would be just as much of a [Fallacies Programmers Think About Currencies](https://gist.github.com/rgs/6509585) as using float dollars/pounds, etc. To support the widest array of currencies, some folks like to use "micros", a concept explained well here by [Sift Science](https://support.siftscience.com/hc/en-us/articles/203869406-The-amount-field). In this case, the new field could easily be called `price_micros`.
+
+Want to keep it simple and go with cents, pence, etc.? Fine, `price_subunits` will do. If somebody grumps about that and you want a more concise name, just call it `amount` and point folks towards that field instead. A thesaurus is handy.
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: I don't understand this. If you're supporting currencies that don't support "unit" and "subunit", then why would use use "price_subunits"?

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+
+Nothing broke for existing use cases, and new functionality was added seamlessly.
+
+> We have too many old fields kicking around, we need to get rid of them.
+
+Deprecations can be communicated in a few ways for API's. For those using OpenAPI v3, you can mark it as `deprecated: true` in the documentation. That's no ideal, of course, as OpenAPI is usually human-readable documentation, sat out of band on a developer portal somewhere. Rarely are OpenAPI schemas shoved into the response like JSON Schema is, so programmatically clients have no real way to access this.
+
+JSON Schema is considering [adding a deprecated keyword](https://github.com/json-schema-org/json-schema-spec/issues/74), and oops I think I'm in charge of making that happen. I'll get back to doing that after this blog post. The idea here would be to pair the schema with a smart SDK (client code) which detects which fields are being used. If the schema marks the `foo` field as deprecated, and the client code then calls `$response->foo`, in PHP that could be snooped on with `function __get` (dynamic programming yaaaay) and fire off deprecation warnings on use.
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Who, there's a lot here. I'd nuke the PHP references (nothing against PHP, sorry not sorry) as it doesn't seem to add much to the discussion. CMIIW, but the point of this paragraph is that if you distribute a client library, you can programatically (at runtime or compile time) make things deprecated.

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+
+All of that said, removing old fields is usually not all that much of a rush or an issue. Over time new developers will be writing new integrations, looking at your new documentation that tells them to use the new field, and your developer newsletters or changelogs just let them know to move away from it over time.
+
+> A carpooling company that has "matches" as a relationship between "drivers" and "passengers", suggesting folks who could ride together, containing properties like "passenger_id" and "driver_id". Now we need to support carpools that can have multiple drivers (i.e: Frank and Sally both take it in turns to drive), so this whole matches concept is garbage.
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: "i.e.," not "i.e:"

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 23]
```diff
@@ -0,0 +1,146 @@
...
+
+[Google Maps are doing this](https://cloud.google.com/maps-platform/user-guide/). 🤷‍♂️
+
+## More Power
+
+The above solutions are a little ad-hoc, and can lead to branched code paths with a bunch of if statements. You can feature flag some of this stuff to help keep things a little tidy, and another approach is to write up change as libraries, something [Stripe refer to as "declarative changes"](https://stripe.com/blog/api-versioning). This approach can be a little heavy handed, but it's something to keep in mind.
+
+## Summary
+
+Evolution involves thinking a little differently to how you approach change. Often there are simple things you can do to keep clients ticking along, and whilst clients will have to change at _some point_, the whole goal here is to allow them a decent amount of time to make that switch, with the minimal change possible during that change, and **no lock-step deploys** required.
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: "*on how" not "to how"

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 24]
```diff
@@ -0,0 +1,146 @@
...
+
+And yes, whilst making a new endpoint to switch `/matches` and `/riders` is essentially the same as `/v1/matches` and `/v2/matches`, you've skipped the [quagmire of tradeoffs](https://blog.apisyouwonthate.com/api-versioning-has-no-right-way-f3c75457c0b7) between global versioning, resource versioning, or **gulp** method versioning. Global versioning has its place, but so does evolution.
+
+Think about it this way. If implementing some change takes twice as long for API developers compared to other versioning approaches, but save 6 or 7 client developer teams from having to do a while bunch of work, testing, etc. to chase new versions, this has been worthwhile to the company in terms of engineering hours spent.
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: *whole bunch of work, not "while bunch of work"

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown [hunk 25]
```diff
@@ -0,0 +1,146 @@
...
+> Deprecating a specific type of authentication from an endpoint, it's time to say goodbye to HTTP Basic Auth.
+
+If the client is making a request with an authorization header, they have some sort of account. If during the signup for that account you've asked them for an email, you can contact them. If you've not got any way to contact them... tough. Monitor how many folks are using HTTP basic, blog about it, shove some videos up, and eventually you're just going to have to turn it off.
+
+The only other approach to helping out here is an SDK. If you slide some deprecation notices into the code months ahead of the cutoff date, you can throw some warnings saying the code is no longer going to work. This gives you a fighting chance for anyone that keeps a bit up to date. For those that don't, you don't have much choice.
+
+Shoving a clear error into your HTTP response (here using the amazing [RFC 7807: Problems for HTTP APIs](https://tools.ietf.org/html/rfc7807)):
+
+```
+HTTP/1.1 403 Forbidden
```
Comment (dplassgit) [inline source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: Does this solution apply to other problems as well?

File: source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown (hunk 21 above)
Comment (philsturgeon) [reply source/2018-05-02-api-evolution-for-rest-http-apis.html.markdown]: > The idea here would be to pair the schema with a smart SDK (client code) which detects which fields are being used. If the schema marks the `foo` field as deprecated, and the client code then calls `$response->foo`, the SDK can raise a deprecation warning. This is achieved by inspecting the schema file at runtime if you offer your schemas in the `Link` header, or at compile time if you're distributing schema files with the SDK. 
Much better thank you!
